### ISSUE TITLE
Filter Rule Query

### ISSUE THREAD

**Issue by gv2870** [basil00/Divert#147]
Hi,

We want to intercept all packets to and from port 80 or port 443, EXCEPT the ones emanating from port 80  or port 443 of local m/c (Local m/c has web server running).

The following rule doesn't compile.  So what rule could be used?

(tcp.PayloadLength > 0) and (tcp.DstPort=80 or tcp.SrcPort=80 or tcp.DstPort=443 or tcp.SrcPort=443) and not ((tcp.SrcPort=80 || tcp.SrcPort=443) and (ip.SrcAddr=192.168.0.143))

Local IP is 192.168.0.143

GR

**Comment by TechnikEmpire** [comment]
If you're diverting traffic through to the local web server as a proxy and back you need to capture outbound only and then send such packets back "inbound" by reversing the src and destination ports. You'd also ignore inbound packets because they'll go to the right place anyway.

If you want to just grab these packets say to sniff I think you'd just do something like (tcp.SrcPort == 80 || tcp.SrcPort == 443) && ip.SrcAddr != 192.168.0.143.

I'll leave the other half of the filter string that handles DstPort equality comparison as an exercise because typing that out on my phone is annoying.

Basically look at the structures of the TCP and IP headers and you can use those properties in the filter string as if you're writing real code. So just write it like an if statement designed to do the same thing.

https://reqrypt.org/windivert-doc.html#filter_language

**Comment by TechnikEmpire** [comment]
I'd have to run that filter string through the WinDivert helper that validates filter strings and let it give me the position of the error and an error message to figure it out. See the docs for those methods if you have time before I do.

**Comment by gv2870** [comment]
It is the position after 'not' in the string
(tcp.PayloadLength > 0) and (tcp.DstPort=80 or tcp.SrcPort=80 or tcp.DstPort=443 or tcp.SrcPort=443) and not  ((tcp.SrcPort=80 || tcp.SrcPort=443) and (ip.SrcAddr=192.168.0.143))
--------^

WINDIVERT_ERROR_UNEXPECTED_TOKEN (Filter expression parse error)
